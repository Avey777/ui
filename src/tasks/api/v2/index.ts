import {Task, TasksApi} from 'src/api'

const getBasePath = () => {
  const host = window.location.hostname
  const port = window.location.port
  const protocol = window.location.protocol

  return `${protocol}//${host}:${port}/api/v2`
}

const createTaskAPI = () => {
  return new TasksApi({basePath: getBasePath()})
}

export const submitNewTask = async (
  organizationId: string,
  flux: string
): Promise<Task> => {
  const api = createTaskAPI()
  const {data} = await api.tasksPost({organizationId, flux})

  return data
}

export const updateTaskFlux = async (id, flux: string): Promise<Task> => {
  const api = createTaskAPI()
  const {data} = await api.tasksTaskIDPatch(id, {flux})

  return data
}

export const updateTaskStatus = async (
  id: string,
  status: Task.StatusEnum
): Promise<Task> => {
  const api = createTaskAPI()
  const {data} = await api.tasksTaskIDPatch(id, {status})

  return data
}

export const getUserTasks = async (user): Promise<Task[]> => {
  const api = createTaskAPI()
  const after = ''
  const {data} = await api.tasksGet(after, user.id)

  return data.tasks
}

export const getTask = async (id): Promise<Task> => {
  const api = createTaskAPI()
  const {data} = await api.tasksTaskIDGet(id)

  return data.task
}

export const deleteTask = (taskID: string) => {
  const api = createTaskAPI()

  return api.tasksTaskIDDelete(taskID)
}
