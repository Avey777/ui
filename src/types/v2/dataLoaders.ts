export enum ConfigurationState {
  Unconfigured = 'unconfigured',
  Verifying = 'verifying',
  Configured = 'configured',
  Loading = 'loading',
  Done = 'done',
  Error = 'error',
}

export enum DataLoaderType {
  CSV = 'CSV',
  Streaming = 'Streaming',
  LineProtocol = 'Line Protocol',
  Empty = '',
}

export interface DataSource {
  name: string
  configured: ConfigurationState
  active: boolean
  configs: any
}
