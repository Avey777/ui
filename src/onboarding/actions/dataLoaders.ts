// Types
import {DataSource, DataLoaderType} from 'src/types/v2/dataLoaders'

export type Action =
  | SetDataLoadersType
  | AddDataSource
  | RemoveDataSource
  | SetActiveDataSource

interface SetDataLoadersType {
  type: 'SET_DATA_LOADERS_TYPE'
  payload: {type: DataLoaderType}
}

export const setDataLoadersType = (
  type: DataLoaderType
): SetDataLoadersType => ({
  type: 'SET_DATA_LOADERS_TYPE',
  payload: {type},
})

interface AddDataSource {
  type: 'ADD_DATA_SOURCE'
  payload: {dataSource: DataSource}
}

export const addDataSource = (dataSource: DataSource): AddDataSource => ({
  type: 'ADD_DATA_SOURCE',
  payload: {dataSource},
})

interface RemoveDataSource {
  type: 'REMOVE_DATA_SOURCE'
  payload: {dataSource: string}
}

export const removeDataSource = (dataSource: string): RemoveDataSource => ({
  type: 'REMOVE_DATA_SOURCE',
  payload: {dataSource},
})

interface SetActiveDataSource {
  type: 'SET_ACTIVE_DATA_SOURCE'
  payload: {dataSource: string}
}

export const setActiveDataSource = (
  dataSource: string
): SetActiveDataSource => ({
  type: 'SET_ACTIVE_DATA_SOURCE',
  payload: {dataSource},
})
