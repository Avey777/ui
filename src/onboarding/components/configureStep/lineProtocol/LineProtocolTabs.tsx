// Libraries
import React, {PureComponent} from 'react'

// Components
import {Radio, ButtonShape} from 'src/clockface'
import DragAndDrop from 'src/shared/components/DragAndDrop'

// Types
import {LineProtocolTab} from 'src/types/v2/dataLoaders'

interface State {
  activeTab: LineProtocolTab
}

interface Props {
  tabs: LineProtocolTab[]
}

class LineProtocolTabs extends PureComponent<Props, State> {
  constructor(props) {
    super(props)
    this.state = {activeTab: this.props.tabs[0]}
  }
  public render() {
    return (
      <>
        {this.tabSelector}
        {this.tabBody}
      </>
    )
  }

  private handleTabClick = (tab: LineProtocolTab) => () => {
    this.setState({activeTab: tab})
  }

  private get tabSelector(): JSX.Element {
    const {tabs} = this.props
    const {activeTab} = this.state
    return (
      <Radio shape={ButtonShape.Default}>
        {tabs.map(t => (
          <Radio.Button
            key={t}
            id={t}
            titleText={t}
            value={t}
            active={activeTab === t}
            onClick={this.handleTabClick(t)}
          >
            {t}
          </Radio.Button>
        ))}
      </Radio>
    )
  }

  private get tabBody(): JSX.Element {
    const {activeTab} = this.state
    if (activeTab === LineProtocolTab.UploadFile) {
      return <DragAndDrop submitText="Upload File" />
    } else {
      return null
    }
  }
}

export default LineProtocolTabs
