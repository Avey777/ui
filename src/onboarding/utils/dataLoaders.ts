// Types
import {DataLoaderType, ConfigurationState} from 'src/types/v2/dataLoaders'

export const getInitialDataSources = (type: DataLoaderType) => {
  if (type === DataLoaderType.Streaming) {
    return []
  }

  return [
    {
      name: type,
      configured: ConfigurationState.Unconfigured,
      active: true,
      configs: null,
    },
  ]
}
