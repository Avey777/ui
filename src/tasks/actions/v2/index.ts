import {AppState} from 'src/types/v2'
import {push} from 'react-router-redux'

import {Task} from 'src/types/v2/tasks'
import {
  submitNewTask,
  updateTaskFlux,
  getUserTasks,
  getTask,
  updateTaskStatus as updateTaskStatusAPI,
  deleteTask as deleteTaskAPI,
} from 'src/tasks/api/v2'
import {getMe} from 'src/shared/apis/v2/user'
import {notify} from 'src/shared/actions/notifications'
import {
  taskNotCreated,
  tasksFetchFailed,
  taskDeleteFailed,
  taskNotFound,
  taskUpdateFailed,
} from 'src/shared/copy/v2/notifications'

import {taskOptionsToFluxScript} from 'src/utils/taskOptionsToFluxScript'

export type Action =
  | SetNewScript
  | SetTasks
  | SetSearchTerm
  | SetCurrentScript
  | SetCurrentTask
  | SetShowInactive
  | SetDropdownOrgID
  | SetTaskInterval
  | SetTaskCron
  | ClearTaskOptions
  | SetTaskOption
  | SetScheduleUnit

type GetStateFunc = () => AppState

export interface ClearTaskOptions {
  type: 'CLEAR_TASK_OPTIONS'
}

export interface SetTaskInterval {
  type: 'SET_TASK_INTERVAL'
  payload: {
    interval: string
  }
}

export interface SetTaskCron {
  type: 'SET_TASK_CRON'
  payload: {
    cron: string
  }
}

export interface SetNewScript {
  type: 'SET_NEW_SCRIPT'
  payload: {
    script: string
  }
}
export interface SetCurrentScript {
  type: 'SET_CURRENT_SCRIPT'
  payload: {
    script: string
  }
}
export interface SetCurrentTask {
  type: 'SET_CURRENT_TASK'
  payload: {
    task: Task
  }
}

export interface SetTasks {
  type: 'SET_TASKS'
  payload: {
    tasks: Task[]
  }
}

export interface SetSearchTerm {
  type: 'SET_SEARCH_TERM'
  payload: {
    searchTerm: string
  }
}

export interface SetShowInactive {
  type: 'SET_SHOW_INACTIVE'
  payload: {}
}

export interface SetDropdownOrgID {
  type: 'SET_DROPDOWN_ORG_ID'
  payload: {
    dropdownOrgID: string
  }
}

export interface SetTaskOption {
  type: 'SET_TASK_OPTION'
  payload: {
    key: string
    value: string
  }
}

export interface SetScheduleUnit {
  type: 'SET_SCHEDULE_UNIT'
  payload: {
    schedule: string
    unit: string
  }
}

export const setTaskOption = (taskOption: {
  key: string
  value: string
}): SetTaskOption => ({
  type: 'SET_TASK_OPTION',
  payload: {...taskOption},
})

export const clearTaskOptions = (): ClearTaskOptions => ({
  type: 'CLEAR_TASK_OPTIONS',
})

export const setNewScript = (script: string): SetNewScript => ({
  type: 'SET_NEW_SCRIPT',
  payload: {script},
})

export const setCurrentScript = (script: string): SetCurrentScript => ({
  type: 'SET_CURRENT_SCRIPT',
  payload: {script},
})

export const setCurrentTask = (task: Task): SetCurrentTask => ({
  type: 'SET_CURRENT_TASK',
  payload: {task},
})

export const setTasks = (tasks: Task[]): SetTasks => ({
  type: 'SET_TASKS',
  payload: {tasks},
})

export const setSearchTerm = (searchTerm: string): SetSearchTerm => ({
  type: 'SET_SEARCH_TERM',
  payload: {searchTerm},
})

export const setShowInactive = (): SetShowInactive => ({
  type: 'SET_SHOW_INACTIVE',
  payload: {},
})

export const setDropdownOrgID = (dropdownOrgID: string): SetDropdownOrgID => ({
  type: 'SET_DROPDOWN_ORG_ID',
  payload: {dropdownOrgID},
})

export const setScheduleUnit = (
  unit: string,
  schedule: string
): SetScheduleUnit => ({
  type: 'SET_SCHEDULE_UNIT',
  payload: {unit, schedule},
})

export const updateTaskStatus = (task: Task) => async (
  dispatch,
  getState: GetStateFunc
) => {
  try {
    const {
      links: {tasks: url},
    } = getState()
    await updateTaskStatusAPI(url, task.id, task.status)

    dispatch(populateTasks())
  } catch (e) {
    console.error(e)
    dispatch(notify(taskDeleteFailed()))
  }
}

export const deleteTask = (task: Task) => async (
  dispatch,
  getState: GetStateFunc
) => {
  try {
    const {
      links: {tasks: url},
    } = getState()

    await deleteTaskAPI(url, task.id)

    dispatch(populateTasks())
  } catch (e) {
    console.error(e)
    dispatch(notify(taskDeleteFailed()))
  }
}

export const populateTasks = () => async (
  dispatch,
  getState: GetStateFunc
): Promise<void> => {
  try {
    const {
      orgs,
      links: {tasks: url, me: meUrl},
    } = getState()

    const user = await getMe(meUrl)
    const tasks = await getUserTasks(url, user)

    const mappedTasks = tasks.map(task => {
      return {
        ...task,
        organization: orgs.find(org => org.id === task.organizationId),
      }
    })

    dispatch(setTasks(mappedTasks))
  } catch (e) {
    console.error(e)
    dispatch(notify(tasksFetchFailed()))
  }
}

export const selectTaskByID = (id: string) => async (
  dispatch,
  getState: GetStateFunc
): Promise<void> => {
  try {
    const {
      links: {tasks: url},
    } = getState()

    const task = await getTask(url, id)

    return dispatch(setCurrentTask(task))
  } catch (e) {
    console.error(e)
    dispatch(goToTasks())
    dispatch(notify(taskNotFound()))
  }
}

export const selectTask = (task: Task) => async dispatch => {
  dispatch(push(`/tasks/${task.id}`))
}

export const goToTasks = () => async dispatch => {
  dispatch(push('/tasks'))
}

export const cancelUpdateTask = () => async dispatch => {
  dispatch(setCurrentTask(null))
  dispatch(goToTasks())
}

export const updateScript = () => async (dispatch, getState: GetStateFunc) => {
  try {
    const {
      links: {tasks: url},
      tasks: {currentScript: script, currentTask: task},
    } = getState()

    await updateTaskFlux(url, task.id, script)

    dispatch(setCurrentTask(null))
    dispatch(goToTasks())
  } catch (e) {
    console.error(e)
    dispatch(notify(taskUpdateFailed()))
  }
}

export const saveNewScript = () => async (
  dispatch,
  getState: GetStateFunc
): Promise<void> => {
  try {
    const {
      orgs,
      links: {tasks: url, me: meUrl},
      tasks: {newScript: script, taskOptions},
    } = await getState()

    const fluxTaskOptions = taskOptionsToFluxScript(taskOptions)
    const scriptWithOptions = `${fluxTaskOptions}\n${script}`

    const user = await getMe(meUrl)

    let org = orgs.find(org => {
      return org.id === taskOptions.orgID
    })

    if (!org) {
      org = orgs[0]
    }

    await submitNewTask(url, user, org, scriptWithOptions)

    dispatch(setNewScript(''))
    dispatch(clearTaskOptions())
    dispatch(goToTasks())
  } catch (e) {
    console.error(e)
    dispatch(notify(taskNotCreated(e.headers['x-influx-error'])))
  }
}
