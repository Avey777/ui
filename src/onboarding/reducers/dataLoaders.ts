//  Utils
import {getInitialDataSources} from 'src/onboarding/utils/dataLoaders'

// Types
import {Action} from 'src/onboarding/actions/dataLoaders'
import {DataSource, DataLoaderType} from 'src/types/v2/dataLoaders'

export interface DataLoadersState {
  dataSources: DataSource[]
  type: DataLoaderType
}

export const INITIAL_STATE: DataLoadersState = {
  dataSources: [],
  type: DataLoaderType.Empty,
}

export default (state = INITIAL_STATE, action: Action): DataLoadersState => {
  switch (action.type) {
    case 'SET_DATA_LOADERS_TYPE':
      return {
        ...state,
        type: action.payload.type,
        dataSources: getInitialDataSources(action.payload.type),
      }
    case 'ADD_DATA_SOURCE':
      return {
        ...state,
        dataSources: [...state.dataSources, action.payload.dataSource],
      }
    case 'REMOVE_DATA_SOURCE':
      return {
        ...state,
        dataSources: state.dataSources.filter(
          ds => ds.name !== action.payload.dataSource
        ),
      }
    case 'SET_ACTIVE_DATA_SOURCE':
      return {
        ...state,
        dataSources: state.dataSources.map(ds => {
          if (ds.name === action.payload.dataSource) {
            return {...ds, active: true}
          }
          return {...ds, active: false}
        }),
      }
    default:
      return state
  }
}
