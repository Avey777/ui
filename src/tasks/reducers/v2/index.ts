import {Action} from 'src/tasks/actions/v2'
import {TaskOptions, TaskSchedule} from 'src/utils/taskOptionsToFluxScript'
import {Task as TaskAPI, User, Organization} from 'src/api'

interface Task extends TaskAPI {
  organization: Organization
  owner?: User
  delay?: string
}
export interface State {
  newScript: string
  currentScript: string
  currentTask?: Task
  tasks: Task[]
  searchTerm: string
  showInactive: boolean
  dropdownOrgID: string
  interval: string
  taskOptions: TaskOptions
}

const defaultTaskOptions: TaskOptions = {
  name: '',
  intervalTime: '1',
  intervalUnit: 'd',
  delayTime: '20',
  delayUnit: 'm',
  cron: '',
  taskScheduleType: TaskSchedule.interval,
  orgID: null,
}

const defaultState: State = {
  newScript: '',
  currentScript: '',
  tasks: [],
  searchTerm: '',
  showInactive: true,
  dropdownOrgID: null,
  interval: '1d',
  taskOptions: defaultTaskOptions,
}

export default (state: State = defaultState, action: Action): State => {
  switch (action.type) {
    case 'CLEAR_TASK_OPTIONS':
      return {
        ...state,
        taskOptions: defaultTaskOptions,
      }
    case 'SET_TASK_OPTION':
      const {key, value} = action.payload

      return {
        ...state,
        taskOptions: {...state.taskOptions, [key]: value},
      }
    case 'SET_SCHEDULE_UNIT':
      const {unit, schedule} = action.payload

      return {
        ...state,
        taskOptions: {...state.taskOptions, [schedule]: unit},
      }
    case 'SET_NEW_SCRIPT':
      return {...state, newScript: action.payload.script}
    case 'SET_CURRENT_SCRIPT':
      return {...state, currentScript: action.payload.script}
    case 'SET_CURRENT_TASK':
      const {task} = action.payload
      let currentScript = ''
      if (task) {
        currentScript = task.flux
      }
      return {...state, currentScript, currentTask: task}
    case 'SET_TASKS':
      return {...state, tasks: action.payload.tasks}
    case 'SET_SEARCH_TERM':
      const {searchTerm} = action.payload
      return {...state, searchTerm}
    case 'SET_SHOW_INACTIVE':
      return {...state, showInactive: !state.showInactive}
    case 'SET_DROPDOWN_ORG_ID':
      const {dropdownOrgID} = action.payload
      return {...state, dropdownOrgID}
    default:
      return state
  }
}
