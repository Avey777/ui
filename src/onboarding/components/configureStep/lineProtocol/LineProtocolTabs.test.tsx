// Libraries
import React from 'react'
import {shallow} from 'enzyme'

// Components
import LineProtocolTabs from './LineProtocolTabs'

import {LineProtocolTab} from 'src/types/v2/dataLoaders'

const setup = (override?) => {
  const props = {
    tabs: [
      LineProtocolTab.UploadFile,
      LineProtocolTab.EnterManually,
      LineProtocolTab.EnterURL,
    ],
    ...override,
  }

  const wrapper = shallow(<LineProtocolTabs {...props} />)

  return {wrapper}
}

describe('LineProtocolTabs', () => {
  describe('rendering', () => {
    it('renders!', () => {
      const {wrapper} = setup()
      expect(wrapper.exists()).toBe(true)

      expect(wrapper).toMatchSnapshot()
    }),
      it('selects first tab as activeTab on load', () => {
        const reorderedTabs = [
          LineProtocolTab.EnterManually,
          LineProtocolTab.UploadFile,
          LineProtocolTab.EnterURL,
        ]
        const {wrapper} = setup({tabs: reorderedTabs})
        expect(wrapper.state('activeTab')).toBe(reorderedTabs[0])
      })
  })
})
