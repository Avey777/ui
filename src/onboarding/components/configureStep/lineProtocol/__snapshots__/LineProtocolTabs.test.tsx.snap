// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LineProtocolTabs rendering renders! 1`] = `
<Fragment>
  <default_1
    onClick={[Function]}
    tabs={
      Array [
        "uploadFile",
        "enterManually",
        "enterURL",
      ]
    }
  />
  <Grid>
    <GridRow>
      <GridColumn
        widthXS={12}
      >
        <div
          className="onboarding--admin-user-form"
        >
          <div
            className="wizard-step--lp-body"
          >
            <default_1
              onURLChange={[Function]}
              urlInput=""
            />
          </div>
          <PrecisionDropdown />
        </div>
      </GridColumn>
    </GridRow>
  </Grid>
</Fragment>
`;
